pkg/framework/didresolver/api.go

```diff
@@ -19,63 +15,22 @@ const (
 	ResolutionResult
 )
 
-// DidMethod operations
+// DidMethod resolves a DID into a result type (default: DidDocumentResult).
+// See the DID resolution spec: https://w3c-ccg.github.io/did-resolution.
 type DidMethod interface {
-	Read(did string, versionID interface{}, versionTime string, noCache bool) ([]byte, error)
+	// Read implements the 'DID Resolution' algorithm defined in
+	// https://w3c-ccg.github.io/did-resolution/#resolving.
+	// Note: at most a single 'Options' will be provided.
+	Read(did string, opts ...Options) ([]byte, error)
+	// Accept registers this DID method resolver with the given method.
 	Accept(method string) bool
 }
 
-// resolveOpts holds the options for did resolve
-type resolveOpts struct {
-	resultType  ResultType
-	versionID   interface{}
-	versionTime string
-	noCache     bool
-}
-
-// ResolveOpt is a did resolve option
-type ResolveOpt func(opts *resolveOpts)
-
-// WithResultType the result type input option can be used to request a certain type of result
-func WithResultType(resultType ResultType) ResolveOpt {
-	return func(opts *resolveOpts) {
-		opts.resultType = resultType
-	}
-}
-
-// WithVersionID the version id input option can be used to request a specific version of a DID Document
-func WithVersionID(versionID interface{}) ResolveOpt {
-	return func(opts *resolveOpts) {
-		opts.versionID = versionID
-	}
-}
-
-// WithVersionTime the version time input option can used to request a specific version of a DID Document
-func WithVersionTime(versionTime time.Time) ResolveOpt {
-	return func(opts *resolveOpts) {
-		opts.versionTime = versionTime.Format(time.RFC3339)
-	}
-}
-
-// WithNoCache the no-cache input option can be used to turn cache on or off
-func WithNoCache(noCache bool) ResolveOpt {
-	return func(opts *resolveOpts) {
-		opts.noCache = noCache
-	}
-}
-
-// didResolverOpts holds the options for resolver instance
-type didResolverOpts struct {
-	didMethods []DidMethod
-}
-
-// Opt is a resolver instance option
-type Opt func(opts *didResolverOpts)
-
-// WithDidMethod to add did method
-// DID methods are checked in the order added
-func WithDidMethod(method DidMethod) Opt {
-	return func(opts *didResolverOpts) {
-		opts.didMethods = append(opts.didMethods, method)
-	}
+// Options holds the optional arguments to the DID resolution algorithm defined in
+// 'Resolving a DID': https://w3c-ccg.github.io/did-resolution/#resolving
+type Options struct {
```

Comment:
I do not understand this change. Why are you changing from functional options to a struct?

Reply:
Two reasons:

* I think we can achieve the same goal (optional args) with less LoC (look at the relevant code that is no longer required with this change (ie. none of those With*** methods)):

```go
resolver.Read(did, didresolver.Options{
    NoCache: true,
    VersionID: 123,
    ResultType: DidDocumentResult,
    })
```
as opposed to:
```go
resolver.Read(did,
    WithNoCache(true),
    WithVersionId(123),
    WithResultType(DidDocumentResult))
```

* Slightly less confusion: there were two sets of "WithXXX" methods, several belonging to `ResolverOpt` and one belonging to`Opts`. The former applied to `didresolver.DIDResolver.Resolve()` while the latter applied to the `didresolver.New()` constructor. I found this a bit confusing and figured it made sense to more clearly distinguish between the two sets by way of their usage. If I hadn't made this change, I would've had to figure out how to make sense of adding `WithXXX` methods as options to `DidMethod.Read()`

Reply:
- WithNoCache should not have taken an argument.
- I’m not sure that WithResultType should even exist externally. It seems that a typed return value method should exist for each type. i.e., different external methods.
- Is 0 the default/initial value for Version ID?

Reply:
Re: Confusion. Yup - there can be some confusion due to the two groups when reading source files.  (I would also suggest better code comments and groupings.)

Note: From an external developer PoV, interestingly GoDoc groups the With functions according to their type. E.g., https://godoc.org/github.com/hyperledger/fabric-sdk-go/pkg/client/resmgmt

See ClientOption vs RequestOption.

Reply:
I also think we could improve this situation by introducing multiple variations of WithVersionID that has an appropriate type as argument rather than interface{}.

(actually that is another advantage of functional options).

Reply:
Opened #127